### PR TITLE
System.Globalization.Extensions cleanup

### PR DESCRIPTION
**Pull request by Clockwork-Muse**
Clean up some tests for System.Globalization.Extensions
Some tests were tightened to better reflect current behavior.

### CODE REVIEW

File: src/System.Globalization.Extensions/tests/IdnMapping/Data/Unicode_6_0/Unicode_6_0_IdnaTest.cs [hunk 1]
```diff
@@ -69,7 +68,10 @@ private static string EscapedToLiteralString(string escaped, int lineNumber)
                 if (i + 1 < escaped.Length && escaped[i] == '\\' && escaped[i + 1] == 'u')
                 {
                     // Verify that the escaped sequence is not malformed
-                    Assert.True(i + 5 < escaped.Length, "There was a problem converting to literal string on Line " + lineNumber);
+                    if (i + 5 >= escaped.Length)
+                    {
+                        throw new ArgumentException($"There was a problem converting to literal string on Line {lineNumber}");
+                    }
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/IdnMapping/Data/Unicode_6_0/Unicode_6_0_IdnaTest.cs]: This one seems fine as an assert.  Why did it change?

File: src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs [hunk 2]
```diff
@@ -83,24 +83,22 @@ public void GetAscii(string unicode, int index, int count, string expected)
             }
             Assert.Equal(expected, new IdnMapping().GetAscii(unicode, index, count));
         }
-        
+
         [Fact]
-        public void TestGetAsciiWithDot()
+        [PlatformSpecific(TestPlatforms.AnyUnix ^ TestPlatforms.OSX)]
+        public void TestGetAsciiWithDot_Succeeds()
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs]: Isn't this ignoring the comment "some versions of Linux succeed and others throw exception"?

File: src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs [hunk 3]
```diff
@@ -83,24 +83,22 @@ public void GetAscii(string unicode, int index, int count, string expected)
             }
             Assert.Equal(expected, new IdnMapping().GetAscii(unicode, index, count));
         }
-        
+
         [Fact]
-        public void TestGetAsciiWithDot()
+        [PlatformSpecific(TestPlatforms.AnyUnix ^ TestPlatforms.OSX)]
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs]: Nit: we've used the logic:
```
TestPlatforms.AnyUnix & ~TestPlatforms.OSX
```
elsewhere.  I realize it's the same thing, but I think this reads a bit more easily, reading at least to me as "any unix and not OSX", whereas I have to stop and think for a moment with the xor.

File: src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs [hunk 4]
```diff
@@ -17,45 +17,51 @@ private string ConvertToString(string codepoints)
...
+            // Windows 8 test data came from http://www.unicode.org/Public/UCD/latest/ucd/NormalizationTest.txt
+            // Windows 7 test came from http://www.unicode.org/Public/3.0-Update1/NormalizationTest-3.0.1.txt
 
             using (Stream stream = typeof(StringNormalizationAllTests).GetTypeInfo().Assembly.GetManifestResourceStream(PlatformDetection.IsWindows7 ? "NormalizationDataWin7" : "NormalizationDataWin8"))
             {
-                Assert.True(stream != null, "Couldn't get the stream for the normalization embedded file");
+                if (stream == null)
+                {
+                    throw new ArgumentNullException("Couldn't get the stream for the normalization embedded file");
+                }
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs]: Another case where I don't see the value of the change.  I'd be fine with changing it to ```Assert.NotNull(stream)```.  ArgumentNullException doesn't seem to make sense here, anyway, as it's not an argument to the function that's null, and the first parameter to ArgumentNullException is supposed to be the parameter name.

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs [hunk 5]
```diff
@@ -24,21 +24,50 @@ public void IsNormalized(string value, NormalizationForm normalizationForm, bool
...
+
+            Assert.Throws<ArgumentException>("strInput", () => "\uFFFE".IsNormalized()); // Invalid codepoint
+            Assert.Throws<ArgumentException>("strInput", () => "\uD800\uD800".IsNormalized()); // Invalid surrogate pair
+
+            Assert.Throws<ArgumentNullException>("strInput", () => StringNormalizationExtensions.IsNormalized(null));
+
             // On desktop IsNormalized is not extension method, trying to do ((string)null).IsNormalized()
             // will get NullReferenceException, in .Net Core we use extension method which will throw
             // ArgumentNullException
-            Assert.True((exception is ArgumentNullException) || (exception is NullReferenceException));
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: @tarekgh, the instance method string.IsNormalized is coming back, right?  Assuming so, @Clockwork-Muse, let's revert this change, as the actual change once that method is available would be to just delete this part of the test that's calling IsNormalized on a null.

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs [hunk 6]
```diff
@@ -63,21 +92,51 @@ public void Normalize(string value, NormalizationForm normalizationForm, string
...
+            Assert.Throws<ArgumentException>("strInput", () => "\uFFFE".Normalize()); // Invalid codepoint
+            Assert.Throws<ArgumentException>("strInput", () => "\uD800\uD800".Normalize()); // Invalid surrogate pair
+
+            Assert.Throws<ArgumentNullException>("strInput", () => StringNormalizationExtensions.Normalize(null));
+
             // On desktop Normalize is not extension method, trying to do ((string)null).Normalize()
             // will get NullReferenceException, in .Net Core we use extension method which will throw
             // ArgumentNullException
-            Assert.True((exception is ArgumentNullException) || (exception is NullReferenceException));
+            Assert.Throws<NullReferenceException>(() => ((string)null).Normalize());
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: Ditto

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs (hunk 5 above)
Comment (tarekgh) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: @stephentoub 

>  the instance method string.IsNormalized is coming back, right?
yes it is already there we can use it now.

File: src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs [hunk 7]
```diff
@@ -83,24 +83,22 @@ public void GetAscii(string unicode, int index, int count, string expected)
...
+        public void TestGetAsciiWithDot_Succeeds()
         {
-            string result = "";
-            Exception ex = Record.Exception(()=> result = new IdnMapping().GetAscii("."));
```
Comment (tarekgh) [inline src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs]: I am not seeing it is worth to split the method into 2 methods. it will be nice to have the test case covering all platforms anyway.

File: src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs [hunk 8]
```diff
@@ -17,45 +17,51 @@ private string ConvertToString(string codepoints)
...
             {
-                sb.Append((char) Int32.Parse(part, NumberStyles.HexNumber | NumberStyles.AllowTrailingWhite | NumberStyles.AllowLeadingWhite, CultureInfo.InvariantCulture));
+                sb.Append((char)Int32.Parse(part, NumberStyles.HexNumber | NumberStyles.AllowTrailingWhite | NumberStyles.AllowLeadingWhite, CultureInfo.InvariantCulture));
             }
             return sb.ToString();
         }
-        
+
         [Fact]
         [ActiveIssue(11803, TestPlatforms.AnyUnix)]
```
Comment (tarekgh) [inline src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs]: > [ActiveIssue(11803, TestPlatforms.AnyUnix)] [](start = 7, length = 44)

looks I already fixed this one but some how missed to enable the test. could you please try to remove this one and look if you see any other failures? I'll try to do it too.

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs (hunk 5 above)
Comment (Clockwork-Muse) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: ...Which is why I'm seeing the results I am.

File: src/System.Globalization.Extensions/tests/IdnMapping/Data/Unicode_6_0/Unicode_6_0_IdnaTest.cs (hunk 1 above)
Comment (tarekgh) [inline src/System.Globalization.Extensions/tests/IdnMapping/Data/Unicode_6_0/Unicode_6_0_IdnaTest.cs]: I am still seeing it is better to keep the assert as the message will be clear if something goes wrong with the test source data. having the test throw exception like that will looks strange in the test reports comparing with the Assert.

Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/IdnMapping/Data/Unicode_6_0/Unicode_6_0_IdnaTest.cs]: Yeah, I'd like to keep these as is.  I understand using exceptions when in non-test-related classes that just happens to be used by tests, as it's good to avoid the Xunit.Assert dependency.  But that's not the case here.  At best I think a change like this just breaks even, and at worst I think it makes the code less understandable.  Assert.* methods, after all, just do some work and generate an exception, so what we're really talking about here is adding code in order to change the exception type, and IMO it's not worth it, including the time it takes to develop and the time it takes to review such changes.

File: src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs (hunk 2 above)
Comment (Clockwork-Muse) [inline src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs]: Yes, it was at the moment, I was just trying to nail it down.  I was assuming that one of the *nixs used in CI would throw, and then I could add that platform to the appropriate side.  They all passed the test, though.

Looking into it a little further, it looks like it was failing on Ubuntu 16(04 at least) and openSUSE, so I'm switching to a ConditionalFact for the moment.

File: src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs (hunk 4 above)
Comment (Clockwork-Muse) [inline src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs]: You're right about the exception being the wrong one.  Actually, since we immediately feed it to a StreamReader, we could probably remove it and just let the reader throw (which will admittedly be an ArgumentNullException...)

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs (hunk 5 above)
Comment (Clockwork-Muse) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: Okay, if it's already there, should the extensions methods be removed?  The ones in common (probably separate PR)?

Comment (tarekgh) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: we cannot remove the extensions methods because of the app comparability. please keep it as it is. thanks.

File: src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs [hunk 9]
```diff
@@ -17,45 +17,51 @@ private string ConvertToString(string codepoints)
...
                     int index = 0;
                     while (!sr.EndOfStream)
                     {
                         string line = sr.ReadLine();
                         string[] parts = line.Split(',');
-                        Assert.True(parts.Length == 5, $"Wrong data at the line {index}");
+                        if (parts.Length != 5)
+                        {
+                            throw new ArgumentException($"Wrong data at the line {index}");
+                        }
```
Comment (Clockwork-Muse) [inline src/System.Globalization.Extensions/tests/Normalization/NormalizationAll.cs]: Hmm, I'd like to avoid just doing the `Assert.Equal(5, parts.Length);`, because that would lose the line number (and make where the error occurred impossible to spot).

**Comment by Clockwork-Muse** [comment]
@stephentoub - Okay, noted.  It wasn't something I was specifically looking for (given it can be hard to find), just something I noticed in passing.

File: src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs [hunk 10]
```diff
@@ -83,24 +83,29 @@ public void GetAscii(string unicode, int index, int count, string expected)
...
+            string result = new IdnMapping().GetAscii(dot);
+            Assert.Equal(dot, result);
+        }
+
+        public static bool GetAsciiWithDotFails =>
+            PlatformDetection.IsWindows ||
+            PlatformDetection.IsOSX ||
+            PlatformDetection.IsUbuntu1604 ||
+            PlatformDetection.IsUbuntu1610 ||
+            RuntimeInformation.IsOSPlatform(OSPlatform.Create("openSUSE"));
```
Comment (tarekgh) [inline src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs]: >RuntimeInformation.IsOSPlatform [](start = 12, length = 31)

nit: could you please add a property in PlatfornDetection (under the common folder) called IsOpenSUSE and then use it here?

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs [hunk 11]
```diff
@@ -65,19 +60,15 @@ public void Normalize(string value, NormalizationForm normalizationForm, string
         [Fact]
         public void Normalize_Invalid()
         {
+            //Assert.True(false, "Framework not skipped");
```
Comment (tarekgh) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: nit: please remove this line

**Comment by tarekgh** [comment]
I put a couple of minor comments, other than that LGTM. thanks @Clockwork-Muse for cleaning up this code.

**Comment by Clockwork-Muse** [comment]
@dotnet-bot help, please

**Comment by Clockwork-Muse** [comment]
... I think things have gotten wedged.

**Comment by stephentoub** [comment]
@dotnet-bot test this please (Jenkins rebooted)

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs [hunk 12]
```diff
@@ -32,13 +32,8 @@ public void IsNormalized_Invalid()
...
             Assert.Throws<ArgumentNullException>("strInput", () => StringNormalizationExtensions.IsNormalized(null));
-            
-            Exception exception = Record.Exception(() => ((string)null).IsNormalized());
-            
-            // On desktop IsNormalized is not extension method, trying to do ((string)null).IsNormalized()
-            // will get NullReferenceException, in .Net Core we use extension method which will throw
-            // ArgumentNullException
-            Assert.True((exception is ArgumentNullException) || (exception is NullReferenceException));
+            // Ensure that extension method not used by current framework.
+            Assert.Throws<NullReferenceException>(() => ((string)null).IsNormalized());
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: I think we can drop this assert entirely: calling an instance method on a null instance produces a NullReferenceException... that's a runtime feature, not something specific to this lib.

File: src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs [hunk 13]
```diff
@@ -71,13 +66,8 @@ public void Normalize_Invalid()
...
             Assert.Throws<ArgumentNullException>("strInput", () => StringNormalizationExtensions.Normalize(null));
-            
-            Exception exception = Record.Exception(() => ((string)null).Normalize());
-            
-            // On desktop Normalize is not extension method, trying to do ((string)null).Normalize()
-            // will get NullReferenceException, in .Net Core we use extension method which will throw
-            // ArgumentNullException
-            Assert.True((exception is ArgumentNullException) || (exception is NullReferenceException));
+            // Ensure that extension method not used by current framework.
+            Assert.Throws<NullReferenceException>(() => ((string)null).Normalize());
```
Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/Normalization/StringNormalizationTests.cs]: Ditto

**Comment by Clockwork-Muse** [comment]
@dotnet-bot test innerloop OpenSUSE13.2 Debug
@dotnet-bot test innerloop OpenSUSE13.2 Release
@dotnet-bot test innerloop Ubuntu16.04 Debug
@dotnet-bot test innerloop Ubuntu16.04 Release
@dotnet-bot test innerloop Ubuntu16.10 Debug
@dotnet-bot test innerloop Ubuntu16.10 Release

( @stephentoub - still need to check if these actually have different behavior, which I was meaning to do)

**Comment by tarekgh** [comment]
@Clockwork-Muse looks OpenSUSE test case is wrong 

```
8:59:03      System.Globalization.Tests.IdnMappingGetAsciiTests.TestGetAsciiWithDot_Succeeds [FAIL]
08:59:03         System.ArgumentException : Decoded string is not a valid IDN name.
08:59:03         Parameter name: unicode
08:59:03         Stack Trace:
08:59:04               at System.Globalization.IdnMapping.GetAsciiCore(Char* unicode, Int32 count)
08:59:04               at System.Globalization.IdnMapping.GetAscii(String unicode, Int32 index, Int32 count)
08:59:04               at System.Globalization.IdnMapping.GetAscii(String unicode, Int32 index)
08:59:04               at System.Globalization.IdnMapping.GetAscii(String unicode)
08:59:04            /mnt/resource/j/workspace/dotnet_corefx/master/opensuse13.2_debug_prtest/src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs(93,0): at System.Globalization.Tests.IdnMappingGetAsciiTests.TestGetAsciiWithDot_Succeeds()
08:59:05   Finished:    System.Globalization.Extensions.Tests
```

could you please fix it and address @stephentoub comments so we'll be able to merge the changes? Thanks.

File: src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs [hunk 14]
```diff
@@ -83,24 +83,29 @@ public void GetAscii(string unicode, int index, int count, string expected)
...
+            string result = new IdnMapping().GetAscii(dot);
+            Assert.Equal(dot, result);
+        }
+
+        public static bool GetAsciiWithDotFails =>
+            PlatformDetection.IsWindows ||
+            PlatformDetection.IsOSX ||
+            PlatformDetection.IsUbuntu1604 ||
+            PlatformDetection.IsUbuntu1610 ||
+            PlatformDetection.IsOpenSUSE;
```
Comment (Clockwork-Muse) [inline src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs]: @stephentoub - Maybe.
The catch with Ubuntu is, 16 fails the call, but 14 succeeds, which is already potentially an issue.  Probably then I should flip it to say "Ubuntu, but not 14".
Also, given #8346 , the behavior of OpenSUSE has changed (maybe important), which we would never have noticed with the original test.

Comment (stephentoub) [inline src/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs]: > maybe important

But if it does change, what are we going to do about it?  Seems like all it would do is cause us to spin our wheels investigating the issue, determining it's just because of an unspecified behavioral difference between versions/distributions, update the test to add it to the list that falls into one behavior or another, etc.  Such detail in the test doesn't end up doing us any favors if it doesn't help us flag real issues in our code and only forces us to do more work.  That's why I'm saying that although the new version is more fine-grained, it doesn't seem warranted, and the old version was better in that regard.

**Comment by Clockwork-Muse** [comment]
@dotnet-bot test innerloop OpenSUSE13.2 Debug
@dotnet-bot test innerloop OpenSUSE13.2 Release
@dotnet-bot test innerloop Ubuntu16.04 Debug
@dotnet-bot test innerloop Ubuntu16.04 Release

**Comment by tarekgh** [comment]
@Clockwork-Muse looks most of Linux based platforms are failing with the same error I mentioned before. looks the platform test is wrong

**Comment by tarekgh** [comment]
test  Innerloop OSX Debug Build and Test please
test Innerloop OSX Release Build and Test please

**Comment by Clockwork-Muse** [comment]
@tarekgh - OSX died for other reasons, the problem is I'm in the dark as to what the proper string for `OSPlatform.Create` is (capitalization wise).  Some of the tests for it use a lowercase "ubuntu", but use an all upper-case "NETBSD".... which is the next thing I was going to try.

**Comment by Clockwork-Muse** [comment]
... I really want that new machine, so I can run a couple of other platforms in VMs, do some of these tests myself.

**Comment by tarekgh** [comment]
thanks @Clockwork-Muse for your effort.

**Comment by Clockwork-Muse** [comment]
@tarekgh - . . . it didn't run the tests for 16.04 or OpenSUSE.  Only the standard tests.

**Comment by Clockwork-Muse** [comment]
@dotnet-bot test innerloop OpenSUSE13.2 Debug
@dotnet-bot test innerloop OpenSUSE13.2 Release
@dotnet-bot test innerloop Ubuntu16.04 Debug
@dotnet-bot test innerloop Ubuntu16.04 Release

**Comment by tarekgh** [comment]
@Clockwork-Muse I didn't notice these didn't run. I don't know why. I am going to revert the changes and let OpenSUSE and Ubuntu 16 tests runs so we can avoid breaking the CI

**Comment by tarekgh** [comment]
I have reverted the change and created another pull request 

https://github.com/dotnet/corefx/pull/12866

**Comment by Clockwork-Muse** [comment]
@tarekgh - I'm assuming they didn't run because they aren't in the default runs.  Perhaps we should have the system keep innerloops requested on a build, just for situations like this?
